be/src/storage/rowset/beta_rowset_writer.cpp

```diff
@@ -425,107 +431,328 @@ Status HorizontalBetaRowsetWriter::_final_merge() {
+
+            auto char_field_indexes = vectorized::ChunkHelper::get_char_field_indexes(schema);
+
+            while (true) {
+                chunk->reset();
+                auto st = itr->get_next(chunk, source_masks.get());
+                if (st.is_end_of_file()) {
+                    break;
+                } else if (st.ok()) {
+                    vectorized::ChunkHelper::padding_char_columns(char_field_indexes, schema, *_context.tablet_schema,
+                                                                  chunk);
+                    _vertical_beta_rowset_writer->add_columns(*chunk, column_groups[i], false);
+                } else {
+                    return st;
+                }
+                if (!source_masks->empty()) {
+                    source_masks->clear();
+                }
+            }
+            itr->close();
+            _vertical_beta_rowset_writer->flush_columns();
         }
-    }
 
-    ChunkIteratorPtr itr = nullptr;
-    // schema change vectorized
-    // schema change with sorting create temporary segment files first
-    // merge them and create final segment files if _context.write_tmp is true
-    if (_context.write_tmp) {
-        if (_context.tablet_schema->keys_type() == KeysType::DUP_KEYS) {
-            itr = new_heap_merge_iterator(seg_iterators);
-        } else if (_context.tablet_schema->keys_type() == KeysType::UNIQUE_KEYS ||
-                   _context.tablet_schema->keys_type() == KeysType::AGG_KEYS) {
-            itr = new_aggregate_iterator(new_heap_merge_iterator(seg_iterators), 0);
-        } else {
-            return Status::NotSupported(fmt::format("HorizontalBetaRowsetWriter not support {} key type final merge",
-                                                    _context.tablet_schema->keys_type()));
+        if (auto st = _vertical_beta_rowset_writer->final_flush(); !st.ok()) {
+            LOG(WARNING) << "failed to final flush rowset when final merge " << _context.tablet_id << ", err=" << st;
+            return st;
         }
-        _context.write_tmp = false;
+        timer.stop();
+        LOG(INFO) << "rowset writer vertical final merge finished. tablet:" << _context.tablet_id
+                  << " #key:" << _context.tablet_schema->num_key_columns() << " input("
+                  << "entry=" << seg_iterators.size() << " rows=" << stats.raw_rows_read
+                  << " bytes=" << PrettyPrinter::print(stats.bytes_read, TUnit::UNIT) << ") output(rows=" << total_rows
+                  << " chunk=" << total_chunk << " bytes=" << PrettyPrinter::print(total_data_size(), TUnit::UNIT)
+                  << ") duration: " << timer.elapsed_time() / 1000000 << "ms";
     } else {
-        itr = new_aggregate_iterator(new_heap_merge_iterator(seg_iterators), 0);
-    }
-    itr->init_encoded_schema(vectorized::EMPTY_GLOBAL_DICTMAPS);
+        auto schema = vectorized::ChunkHelper::convert_schema_to_format_v2(*_context.tablet_schema);
```

Comment:
I have not found many differences between the if/else and so many redundant codes.